sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotHideBaseClassMethods.cs

```diff
@@ -0,0 +1,179 @@
+                    }
+
+                    classTypeSymbol = classTypeSymbol.BaseType;
+                }
+            }
+
+            public Diagnostic FindIssue(MemberDeclarationSyntax memberDeclaration)
+            {
+                var methodSymbol = semanticModel.GetDeclaredSymbol(memberDeclaration) as IMethodSymbol;
+                var issueLocation = (memberDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+
+                if (methodSymbol == null || issueLocation == null)
+                {
+                    return null;
+                }
+
+                var baseMethodHidden = FindBaseMethodHiddenByMethod(methodSymbol);
+                return baseMethodHidden != null ? Diagnostic.Create(rule, issueLocation, baseMethodHidden) : null;
+            }
+
+            private IMethodSymbol FindBaseMethodHiddenByMethod(IMethodSymbol methodSymbol)
+            {
+                var baseMemberCandidates = allBaseClassMethods.Where(m => m.Name == methodSymbol.Name);
+
+                IMethodSymbol hiddenBaseMethodCandidate = null;
+                var hasBaseMethodWithSameSignature = baseMemberCandidates.Any(baseMember =>
+                {
+                    var result = ComputSignatureMatch(baseMember, methodSymbol);
+                    if (result == Match.WeaklyDerived)
+                    {
+                        hiddenBaseMethodCandidate = hiddenBaseMethodCandidate ?? baseMember;
+                    }
+
+                    return result == Match.Identical;
+                });
+
+                return hasBaseMethodWithSameSignature ? null : hiddenBaseMethodCandidate;
+            }
+
+            private enum Match { Different, Identical, WeaklyDerived }
+
+            private Match ComputSignatureMatch(IMethodSymbol baseMethodSymbol, IMethodSymbol methodSymbol)
+            {
+                var baseMethodParams = baseMethodSymbol.Parameters;
+                var methodParams = methodSymbol.Parameters;
+
+                if (baseMethodParams.Length != methodParams.Length)
+                {
+                    return Match.Different;
+                }
+
+                bool hasWeaklyDerivedParams = false;
+                for (int i = 0; i < methodParams.Length; i++)
+                {
+                    var match = ComputeParameterMatch(baseMethodParams[i], methodParams[i]);
+
+                    if (match == Match.Different)
+                    {
+                        return Match.Different;
+                    }
+
+                    if (match == Match.WeaklyDerived)
+                    {
+                        hasWeaklyDerivedParams = true;
+                    }
+                }
+
+                return hasWeaklyDerivedParams ? Match.WeaklyDerived : Match.Identical;
+            }
+
+            private Match ComputeParameterMatch(IParameterSymbol baseParam, IParameterSymbol methodParam)
+            {
+                if (baseParam.Type.TypeKind == TypeKind.TypeParameter)
```

Comment:
Not sure what more it is doing but I would use `baseParam.Type.TypeKind.Is(TypeKind.TypeParameter)`.